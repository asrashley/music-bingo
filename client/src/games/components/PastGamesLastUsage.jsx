import React, { useReducer, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Table, Column, HeaderCell } from 'rsuite-table';
import { reverse } from 'named-urls';

import { DateTimeCell, ElapsedTimeCell, LinkCell } from '../../components';
import { PastGamesThemePropType } from '../types/PastGamesThemeFields';

import { routes } from '../../routes/routes';

/* MonthDetail is a map of theme key to game count */
export const MonthDetail = PropTypes.objectOf(PropTypes.number);

export function sortData(data, sortColumn, sortType) {
    const result = [...data];
    result.sort((a, b) => {
        if (a[sortColumn] < b[sortColumn]) {
            return (sortType === "desc") ? -1 : 1;
        }
        if (a[sortColumn] > b[sortColumn]) {
            return (sortType === "desc") ? 1 : -1;
        }
        return 0;
    });
    return result;
}

function reducer(state, action) {
    switch (action.type) {
        case 'setData':
            return {
                ...state,
                data: action.data,
            };
        case 'setSort': {
            const { data, sortColumn, sortType } = action;
            return {
                ...state,
                sortColumn,
                sortType,
                data: sortData(data, sortColumn, sortType),
            };
        }
        default:
            return state;
    }
}

export function PastGamesLastUsage({ loading, themes }) {
    const [state, dispatch] = useReducer(reducer, {
        sortColumn: 'theme',
        sortType: 'desc',
        data: themes ?? [],
    });
    const { sortColumn, sortType, data } = state;

    useEffect(() => {
        dispatch({ type: 'setData', data: themes });
    }, [themes]);

    const pastGamesLink = (theme) => {
        return reverse(`${routes.pastGamesByTheme}`, { slug: theme.slug });
    }

    return (<Table
        className="past-games-calendar table table-striped"
        title="Theme History"
        rowKey='slug'
        autoHeight
        bordered
        cellBordered
        loading={loading}
        data={data}
        sortColumn={sortColumn}
        sortType={sortType}
        onSortColumn={(sortColumn, sortType) => dispatch({ type: 'setSort', sortColumn, sortType, data })}
        width={350 + 400}
    >
        <Column width={350} align="left" sortable fixed resizable>
            <HeaderCell>Theme</HeaderCell>
            <LinkCell dataKey="title" to={pastGamesLink} />
        </Column>
        <Column width={200} align="right" sortable fixed>
            <HeaderCell>Last Used</HeaderCell>
            <DateTimeCell dataKey="lastUsed" />
        </Column>
        <Column width={200} align="right" sortable fixed>
            <HeaderCell>Elapsed Time</HeaderCell>
            <ElapsedTimeCell dataKey="elapsedTime" />
        </Column>
    </Table>);
}
PastGamesLastUsage.propTypes = {
    themes: PropTypes.arrayOf(PastGamesThemePropType),
    loading: PropTypes.bool.isRequired,
};
