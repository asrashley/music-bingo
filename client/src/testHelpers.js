import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

