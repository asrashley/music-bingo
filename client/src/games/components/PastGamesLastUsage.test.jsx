import { within, fireEvent } from '@testing-library/react';
import { renderWithProviders } from '../../testHelpers';
import { PastGamesLastUsage, sortData } from './PastGamesLastUsage';

const themes = [{
    "key": "100nows",
    "title": "100 Nows",
    "row": {
        "2020-02": 1,
        "2020-08": 1,
        "2022-04": 1,
        "2022-11": 1,
        "2023-09": 1
    },
    "lastUsed": new Date("2023-09-21T17:52:25.041Z"),
    "elapsedTime": 5795103227
}, {
    "key": "1973",
    "title": "1973",
    "row": {
        "2023-10": 1
    },
    "lastUsed": new Date("2023-10-05T16:44:53.615Z"),
    "elapsedTime": 4589554653
}, {
    "key": "2000s",
    "title": "2000s",
    "row": {
        "2018-08": 1,
        "2019-06": 1,
        "2020-01": 1,
        "2021-11": 1,
        "2022-08": 1,
        "2023-05": 1
    },
    "lastUsed": new Date("2023-05-18T09:49:39.188Z"),
    "elapsedTime": 16710469080
}];

describe('PastGamesLastUsage component', () => {
    beforeAll(() => {
        jest.useFakeTimers('modern');
        jest.setSystemTime(new Date('08 Feb 2023 10:12:00 GMT').getTime());
    });

    afterAll(() => jest.useRealTimers());

    it('PastGamesLastUsage matches snapshot', () => {
        const { asFragment } = renderWithProviders(<PastGamesLastUsage themes={themes} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it.each(['title', 'lastUsed', 'elapsedTime'])('PastGamesLastUsage can sort by %s', async (column) => {
        function clickSortButton() {
            if (column === 'title') {
                fireEvent.click(getByText('Theme'));
            } else if (column === 'lastUsed') {
                fireEvent.click(getByText('Last Used'));
            } else {
                fireEvent.click(getByText('Elapsed Time'));
            }
        }
        const { getByText, getAllBySelector, findAllBySelector } = renderWithProviders(<PastGamesLastUsage themes={themes} />);
        getByText('100 Nows');
        let expected = themes.map(item => item.title);
        let tableRows = getAllBySelector('.rs-table-body-row-wrapper .rs-table-row');
        tableRows.forEach((row, index) => {
            within(row).getByText(expected[index]);
        });

        let sortedThemes = sortData(themes, column, 'desc');
        expected = sortedThemes.map(item => item.title);
        clickSortButton();
        tableRows = await findAllBySelector('.rs-table-body-row-wrapper .rs-table-row');
        tableRows.forEach((row, index) => {
            within(row).getByText(expected[index]);
        });

        sortedThemes = sortData(sortedThemes, column, 'asc');
        expected = sortedThemes.map(item => item.title);
        clickSortButton();
        tableRows = await findAllBySelector('.rs-table-body-row-wrapper .rs-table-row');
        tableRows.forEach((row, index) => {
            within(row).getByText(expected[index]);
        });
    });
});