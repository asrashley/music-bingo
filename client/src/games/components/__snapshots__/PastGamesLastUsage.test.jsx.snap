// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PastGamesLastUsage component > PastGamesLastUsage matches snapshot when loaded 1`] = `
<DocumentFragment>
  <div
    aria-busy="false"
    aria-colcount="3"
    aria-rowcount="4"
    class="past-games-calendar table table-striped rs-table rs-table-bordered rs-table-hover rs-table-cell-bordered"
    role="grid"
    style="width: 750px; height: 184px;"
    tabindex="-1"
    title="Theme History"
  >
    <div
      class="rs-table-header-row-wrapper"
      role="rowgroup"
    >
      <div
        aria-rowindex="1"
        class="rs-table-row rs-table-row-header"
        role="row"
        style="min-width: 750px; height: 40px; transform: translate(0px,0px);"
      >
        <div
          class="rs-table-cell-group rs-table-cell-group-scroll"
          style="transform: translate(0px,0px);"
        >
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="1"
              class="rs-table-cell rs-table-cell-first"
              role="columnheader"
              style="width: 350px; height: 40px; z-index: 0; left: 0px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-start; width: 350px; height: 40px;"
              >
                Theme
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
            <div
              class="rs-table-column-resize-spanner"
              role="button"
              style="left: 348px; height: 39px;"
              tabindex="-1"
            />
          </div>
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="2"
              class="rs-table-cell"
              role="columnheader"
              style="width: 200px; height: 40px; z-index: 0; left: 350px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 40px;"
              >
                Last Used
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="3"
              class="rs-table-cell rs-table-cell-last"
              role="columnheader"
              style="width: 200px; height: 40px; z-index: 0; left: 550px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 40px;"
              >
                Elapsed Time
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-table-body-row-wrapper"
      role="rowgroup"
      style="top: 40px; height: 144px;"
    >
      <div
        class="rs-table-body-wheel-area"
        style="position: absolute; height: 138px; min-height: 184px;"
      >
        <div
          aria-rowindex="2"
          class="rs-table-row"
          role="row"
          style="min-width: 750px; height: 46px; transform: translate(0px,0px);"
        >
          <div
            class="rs-table-cell-group rs-table-cell-group-scroll"
            style="transform: translate(0px,0px);"
          >
            <div
              aria-colindex="1"
              class="rs-table-cell rs-table-cell-first"
              role="gridcell"
              style="width: 350px; height: 46px; z-index: 0; left: 0px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-start; width: 350px; height: 46px;"
              >
                <span
                  class=""
                >
                  <a
                    href="/history/themes/100nows"
                  >
                    100 Nows
                  </a>
                </span>
              </div>
            </div>
            <div
              aria-colindex="2"
              class="rs-table-cell"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 350px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  18:52:25 21/09/2023
                </span>
              </div>
            </div>
            <div
              aria-colindex="3"
              class="rs-table-cell rs-table-cell-last"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 550px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  <span
                    class="elapsed-time"
                  >
                    9 weeks ago
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-rowindex="3"
          class="rs-table-row"
          role="row"
          style="min-width: 750px; height: 46px; transform: translate(0px,46px);"
        >
          <div
            class="rs-table-cell-group rs-table-cell-group-scroll"
            style="transform: translate(0px,0px);"
          >
            <div
              aria-colindex="1"
              class="rs-table-cell rs-table-cell-first"
              role="gridcell"
              style="width: 350px; height: 46px; z-index: 0; left: 0px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-start; width: 350px; height: 46px;"
              >
                <span
                  class=""
                >
                  <a
                    href="/history/themes/1973"
                  >
                    1973
                  </a>
                </span>
              </div>
            </div>
            <div
              aria-colindex="2"
              class="rs-table-cell"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 350px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  17:44:53 05/10/2023
                </span>
              </div>
            </div>
            <div
              aria-colindex="3"
              class="rs-table-cell rs-table-cell-last"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 550px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  <span
                    class="elapsed-time"
                  >
                    7 weeks ago
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-rowindex="4"
          class="rs-table-row"
          role="row"
          style="min-width: 750px; height: 46px; transform: translate(0px,92px);"
        >
          <div
            class="rs-table-cell-group rs-table-cell-group-scroll"
            style="transform: translate(0px,0px);"
          >
            <div
              aria-colindex="1"
              class="rs-table-cell rs-table-cell-first"
              role="gridcell"
              style="width: 350px; height: 46px; z-index: 0; left: 0px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-start; width: 350px; height: 46px;"
              >
                <span
                  class=""
                >
                  <a
                    href="/history/themes/2000s"
                  >
                    2000s
                  </a>
                </span>
              </div>
            </div>
            <div
              aria-colindex="2"
              class="rs-table-cell"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 350px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  10:49:39 18/05/2023
                </span>
              </div>
            </div>
            <div
              aria-colindex="3"
              class="rs-table-cell rs-table-cell-last"
              role="gridcell"
              style="width: 200px; height: 46px; z-index: 0; left: 550px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 46px;"
              >
                <span
                  class=""
                >
                  <span
                    class="elapsed-time"
                  >
                    27 weeks ago
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-table-mouse-area"
      style="height: 184px;"
    >
      <span
        style="height: 39px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PastGamesLastUsage component > PastGamesLastUsage matches snapshot when loading 1`] = `
<DocumentFragment>
  <div
    aria-busy="true"
    aria-colcount="3"
    aria-rowcount="1"
    class="past-games-calendar table table-striped rs-table rs-table-bordered rs-table-hover rs-table-loading rs-table-cell-bordered"
    role="grid"
    style="width: 750px; height: 200px;"
    tabindex="-1"
    title="Theme History"
  >
    <div
      class="rs-table-header-row-wrapper"
      role="rowgroup"
    >
      <div
        aria-rowindex="1"
        class="rs-table-row rs-table-row-header"
        role="row"
        style="min-width: 750px; height: 40px; transform: translate(0px,0px);"
      >
        <div
          class="rs-table-cell-group rs-table-cell-group-scroll"
          style="transform: translate(0px,0px);"
        >
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="1"
              class="rs-table-cell rs-table-cell-first"
              role="columnheader"
              style="width: 350px; height: 40px; z-index: 0; left: 0px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-start; width: 350px; height: 40px;"
              >
                Theme
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
            <div
              class="rs-table-column-resize-spanner"
              role="button"
              style="left: 348px; height: 39px;"
              tabindex="-1"
            />
          </div>
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="2"
              class="rs-table-cell"
              role="columnheader"
              style="width: 200px; height: 40px; z-index: 0; left: 350px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 40px;"
              >
                Last Used
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="rs-table-cell-header rs-table-cell-header-sortable"
          >
            <div
              aria-colindex="3"
              class="rs-table-cell rs-table-cell-last"
              role="columnheader"
              style="width: 200px; height: 40px; z-index: 0; left: 550px;"
            >
              <div
                class="rs-table-cell-content"
                style="display: flex; flex-wrap: wrap; justify-content: flex-end; width: 200px; height: 40px;"
              >
                Elapsed Time
                <span
                  class="rs-table-cell-header-sort-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="sort"
                    class="rs-table-cell-header-icon-sort rs-icon"
                    data-category="direction"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 16 16"
                    width="1em"
                  >
                    <path
                      d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-table-body-row-wrapper"
      role="rowgroup"
      style="top: 40px; height: 160px;"
    >
      <div
        class="rs-table-body-wheel-area"
        style="position: absolute; height: 0px; min-height: 200px;"
      />
      <div
        class="rs-table-loader-wrapper"
      >
        <div
          class="rs-table-loader"
        >
          <i
            class="rs-table-loader-icon"
          />
          <span
            class="rs-table-loader-text"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
    <div
      class="rs-table-mouse-area"
      style="height: 200px;"
    >
      <span
        style="height: 39px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
